BUG: special: Fix behavior of `gamma` and `gammasgn` at poles of the Gamma function

steppi opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes gh-19071, gh-21810

#### What does this implement/fix?
<!--Please explain your changes.-->
This PR fixes the behavior of `special.gamma` and `special.gammasgn` at poles of the `gamma` function. As described in #21810, we currently have `special.gamma(n) == np.inf` for  every non-positive integer `n`, and bizarrely `special.gamma(-np.inf) == -np.inf`.  As described in #19710, `gammasgn(n) == 0.0` for every non-positive integer `n` and even more bizarrely, `gammasgn(-np.inf) == 1.0`.

The sign of $\Gamma(x)$ at each pole depends on the direction in which $x$ approaches the pole, so it doesn't make sense to return $+\infty$ at all poles. Accordingly, the [C99 standard](https://en.cppreference.com/w/c/numeric/math/tgamma) recommends the following behavior for the gamma function in special cases.

1. If the argument is $\pm 0$, $\pm \infty$ is returned.
2. If the argument is a negative integer, NaN is returned.
3. If the argument is $-\infty$, NaN is returned.
4. If the argument is $+\infty$, $+\infty$ is returned
5. If the argument is NaN, NaN is returned.

I've updated `special.gamma` to follow this standard. Note that signed zeros offer the ability to specify the sign of infinity of `gamma(x)` at `x == 0`. I've also updated `gammasgn` to be consistent with the above.

 Note that as described in #21810, the original cephes gamma function implementation gets these cases right, and `special.gamma`'s current behavior was a deliberate choice. I had speculated earlier that `gamma(x) == +inf` at poles may have been enforced so that `gamma` ratios can be computed naively without worrying about hitting a pole in the denominator. This seems to be born out, after making the above changes, I found that the `cephes` implementations of `beta` and `betaln` needed to be updated because they rely on this behavior. In addition, `test_pbvd_points` needed to be updated as well because it divides by `gamma` in the computation of reference values.

I think the preferred thing to do when computing gamma ratios should be to use `rgamma` for `gamma`s in the denominators, but I discovered that `rgamma(x)` can be much less accurate than `1.0/gamma(x)` for `x` away from 0. This caused test failures in `test_mpmath` for `betaln` when I switched `betaln`'s implementation to use `rgamma` internally. I fixed this by updating the scalar kernel of `rgamma` to just compute `1.0/gamma(x)` away from the origin, with special handling at the poles. 




The strange behavior of `gammasgn` at poles is currently enforced through a test which compares `special.gammasgn(x)` with `np.sign(special.rgamma(x)`.

https://github.com/scipy/scipy/blob/0f6de5836a9fa7bdd5121810ccd452bdde5295db/scipy/special/tests/test_basic.py#L400-L402

and since `rgamma(x) == 0` at poles, the test passes when `gammasgn(x) == 0` at poles. I'm not sure if there was a reason for deciding on this behavior. I've simply changed the test cases.

I've also added additional tests for `gamma` at poles and removed a test case enforcing `gamma(-1) == +inf`.

#### Additional information

The details section here contains the justification for choosing a cutoff for evaluating `rgamma` as `1/gamma` rather than following the current `rgamma` implementation.

<details>

Here's a plot of the log10 of relative error against an mpmath reference for both `rgamma` and `1/gamma` over the range `(-99.9, 99.9)` showing how bad things can be

![rgamma_inaccuracy_wide](https://github.com/user-attachments/assets/8583fd59-235a-4686-98f0-5eb7b0335561)

Here are the deciles for relative error

| method   |   min |   10% |         20% |         30% |         40% |         50% |         60% |         70% |         80% |         90% |         max |
|----------|-------|-------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
| rgamma   |     0 |     0 | 1.68305e-16 | 3.08089e-16 | 2.56746e-15 | 7.53686e-15 | 1.32971e-14 | 1.95882e-14 | 2.8148e-14  | 4.31268e-14 | 1.35586e-13 |
| 1/gamma  |     0 |     0 | 0           | 0           | 1.22212e-16 | 1.43391e-16 | 1.67001e-16 | 1.92162e-16 | 2.24835e-16 | 3.22916e-16 | 1.07467e-15 |

zooming in, it looks like `rgamma` starts to do a little better near the origin

![rgaa](https://github.com/user-attachments/assets/1dc90841-a2b6-4ae3-9321-8f7bc5b2db75)

I eyeballed this plot, and chose a cutoff of `|x| < 4` for using the existing `rgamma`. Here's the deciles for relative error just for `|x| < 4`.

| method   |   min |   10% |   20% |   30% |   40% |   50% |         60% |         70% |         80% |        90% |         max |
|----------|-------|-------|-------|-------|-------|-------|-------------|-------------|-------------|------------|-------------|
| rgamma   |     0 |     0 |     0 |     0 |     0 |     0 | 0           | 1.26661e-16 | 1.58249e-16 | 2.0817e-16 | 3.13087e-16 |
| 1/gamma  |     0 |     0 |     0 |     0 |     0 |     0 | 1.26577e-16 | 1.51009e-16 | 1.92822e-16 | 2.2071e-16 | 4.24229e-16 |

I used the following function to calculate these plots and tables

```python
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import scipy.special as special

from mpmath import mp

def get_rgamma_table_and_plot(x):
    res1 = special.rgamma(x)
    res2 = 1.0 / special.gamma(x)
    reference = np.fromiter((float(mp.rgamma(t)) for t in x), dtype=np.float64)

    rel_err1 = abs(res1 - reference)/abs(reference)
    rel_err2 = abs(res2 - reference)/abs(reference)

    # rgamma deciles
    dec1 = np.quantile(rel_err1, np.linspace(0, 1, 11))
    # gamma deciles
    dec2 = np.quantile(rel_err2, np.linspace(0, 1, 11))

    table = pd.DataFrame(
        [
            ["rgamma"] + dec1.tolist(),
            ["1/gamma"] + dec2.tolist(),
        ],
        columns=["method", "min"] + [f"{n}%" for n in range(10, 91, 10)] + ["max"],
    ).to_markdown(index=False, tablefmt="github")

    fig, ax = plt.subplots()
    ax.plot(x, np.where(rel_err1 == 0, -16, np.log10(rel_err1)), label="rgamma")
    ax.plot(x, np.where(rel_err2 == 0, -16, np.log10(rel_err2)), label="1/gamma")
    ax.set_xlabel("x")
    ax.set_ylabel("log10(rel error)")
    ax.legend()
    return table, fig
```

</details>

##### Potential impact

I ran the command

```
ack Gamma --type=cc
```

from within the `scipy/scipy/special`

directory to identify places where `Gamma` is used to see if there are other places where it appears in a denominator. In most of the examples, changing the pole behavior in this PR won't matter, because these special cases are already handled. To make it so the reviewer doesn't have to check that these cases are already handled in each of the instances where `gamma` appears in the denominator,  I've decided to uniformly switch to `rgamma` whenever `gamma` appears in the denominator. The exception is when you have a product with a `gamma` function in the denominator. In that case, typically there was a specific choice made to multiply first before dividing, typically in order to get better numerical accuracy. I'll enumerate these cases below, and will work on verifying no issues will result from changing the pole behavior.

##### `besselpoly`

https://github.com/scipy/scipy/blob/d1feb94fb2d8bdd845c79d276c9ca027f86e1841/scipy/special/xsf/cephes/besselpoly.h#L31-L36

but we see that these cases are already handled, so it's not an issue.

##### real valued `hyp2f1`:

https://github.com/scipy/scipy/blob/d1feb94fb2d8bdd845c79d276c9ca027f86e1841/scipy/special/xsf/cephes/hyp2f1.h#L274-L276

This case above occurs within a branch

```C++
            if (x > 0.9 && !(neg_int_a || neg_int_b)) {
```

so `a` and `b` cannot be poles. The name `neg_int_a` is not quite right, it really means non-positive.  From lines 168-174

```C++
            if (a <= 0 && std::abs(a - ia) < hyp2f1_EPS) { /* a is a negative integer */
                neg_int_a = 1;
            }

            if (b <= 0 && std::abs(b - ib) < hyp2f1_EPS) { /* b is a negative integer */
                neg_int_b = 1;
            }
```

https://github.com/scipy/scipy/blob/d1feb94fb2d8bdd845c79d276c9ca027f86e1841/scipy/special/xsf/cephes/hyp2f1.h#L295-L298

https://github.com/scipy/scipy/blob/d1feb94fb2d8bdd845c79d276c9ca027f86e1841/scipy/special/xsf/cephes/hyp2f1.h#L496-L497

https://github.com/scipy/scipy/blob/d1feb94fb2d8bdd845c79d276c9ca027f86e1841/scipy/special/xsf/cephes/hyp2f1.h#L535


##### complex valued `hyp2f1`

https://github.com/scipy/scipy/blob/d1feb94fb2d8bdd845c79d276c9ca027f86e1841/scipy/special/xsf/hyp2f1.h#L210

https://github.com/scipy/scipy/blob/d1feb94fb2d8bdd845c79d276c9ca027f86e1841/scipy/special/xsf/hyp2f1.h#L477C13-L479

https://github.com/scipy/scipy/blob/d1feb94fb2d8bdd845c79d276c9ca027f86e1841/scipy/special/xsf/hyp2f1.h#L489-L490

https://github.com/scipy/scipy/blob/d1feb94fb2d8bdd845c79d276c9ca027f86e1841/scipy/special/xsf/hyp2f1.h#L506


#### Some thoughts

Seeing  cases where we rely on `gamma(n) == inf` for `n` a non-positive integer at least makes me realize why this convention was chosen. It does give a shortcut when evaluating gamma ratios, and for cases like

```python
gamma(a)*gamma(b)/(gamma(c)*gamma(d))
```

switching to using `rgamma` can result in reduced accuracy. I think it's better to follow the standard and most mathematically correct here way, but since we relied on this behavior internally, maybe there should be a deprecation period, or perhaps we should even keep a `kwarg` around permanently to switch between following the standard and returning `+inf` at the poles, changing the default after a deprecation period. If no one else sees an issue with it, I'm fine with treating this as a bug and just fixing it at once though.


<!--Any additional information you think is important.-->

steppi committed:
BUG: fix behavior at poles for gamma function

steppi committed:
MAINT: Remove use of goto for gamma pole handling

steppi committed:
TST: Add tests for gamma at the poles

steppi committed:
TST: Remove incorrect test case

steppi committed:
MAINT: Use rgamma in beta for gamma ratios

steppi committed:
MAINT: Improve rgamma

steppi committed:
TST: Use rgamma instead of divide by gamma in test case

steppi committed:
BUG: Fix gammasgn behavior

steppi committed:
TST: Add correct test cases for gammasgn

steppi committed:
TST: add gammasgn test cases for infinities

steppi committed:
BUG: Handle negative inf properly in gammasgn

steppi committed:
MAINT: Adjust cutoff for using 1/gamma in rgamma

steppi committed:
MAINT: Replace divisions by gamma with rgamma

steppi committed:
MAINT: Prevent over-aggressive optimizations on 32bit linux

h-vetinari commented on scipy/special/xsf/cephes/gamma.h:
```diff
@@ -360,16 +375,22 @@ namespace cephes {
+	if (x - fx == 0.0) {
+	    return std::numeric_limits<double>::quiet_NaN();
+	}
+	if (static_cast<int>(fx) % 2) {
+	    return -1.0;
+	}
```
I'd add a comment à la: "sign of the gamma function for x in (-N,-N+1) is (-1)^N"

h-vetinari reviewed (commented):
Can't really judge the implementation change of `rgamma` from 

```C
        if (x > 34.84425627277176174) {
            return std::exp(-lgam(x));
        }
```
to
```C
	if (std::abs(x) > 4.0) {
	    return 1.0 / Gamma(x);
	}
```
but aside from a minor C&P error in the tests, this looks good to me. Thanks for tackling this! 🙏

steppi commented on scipy/special/tests/test_basic.py:
```diff
@@ -2715,9 +2720,26 @@ def test_rgamma(self):
+        [
+            # infinities
+            ([-np.inf, np.inf], [np.nan, np.inf]),
+            # zeros
+            ([-0.0, 0.0], [-np.inf, np.inf]),
```
Ah, good point about "zeros".  I meant two "zeros" for input as in `+0.0` and `-0.0`, but then it makes it sound like I'm talking about zeros of the zeroless gamma function. I'll change the comment to something like `# positive and negative zero`

The second pair here isn't actually repeated by the way, the format is that it's a tuple, of input value `x`, and the expected reference value. It's not obvious though. Because these are annoying to read due to the context for what each element of the tuple means (here `"x,expected"`) being far from the tuples, I actually used named tuples for the `hyp2f1` test suite:

https://github.com/scipy/scipy/blob/3367349b6bea4742c0ab751ba1a624303707707a/scipy/special/tests/test_hyp2f1.py#L79-L85

which makes it a little more clear what's going on. I could do that here, but it might be obstrusive to add to `test_basic.py`.

h-vetinari replied:
> The second pair here isn't actually repeated by the way, the format is that it's a tuple, of input value `x`, and the expected reference value. It's not obvious though.

Indeed I missed that both `x` and `expected` are lists here.

steppi commented:
Thanks @h-vetinari! For judging the new `rgamma` implementation click on Details in the  **Additional Information** section above. It will expand and show tables and plots demonstrating the new implementation is superior, along with the function i used to generate the tables and plots.

![image](https://github.com/user-attachments/assets/630baa4d-d533-470a-9062-b8a0621be074)

h-vetinari commented:
Sorry I missed that part 😅

Sounds great!

steppi commented:
Something that's bothering me is that prior to this PR, `rgamma` is actually fairly inaccurate, so we have a situation where we advise those using `scipy>=1.15` to use `rgamma` in gamma ratios, but for `scipy<1.15` using `rgamma` may be unacceptably inaccurate. @h-vetinari, do you have any thoughts on this?

steppi committed:
MAINT: respond to review comments

steppi committed:
DOC: Add note to gamma docstring about behavior at poles [docs only]

steppi committed:
DOC: Fix bug in docstring citation syntax [docs only]

steppi committed:
TST: Add test case failing with c99 conforming gamma pole behavior

steppi committed:
TST: Fix hyp2f1 test_miscellaneous

steppi committed:
MAINT: Remove gamma's in denominators in hyp2f1

steppi commented:
By tracing through the cephes hyp2f1 source code I was able to find an example input that returns NaN after fixing pole behavior of the gamma function. I added this example as a test case. To fix it, and avoid having to trace through other functions, I simply replaced all instances having a `gamma` in the denominator with `rgamma`.

I also added a note to the docstring explaining this change.

steppi commented:
I have extensive test cases for complex `hyp2f1` through `scipy/special/_precompute/hyp2f1_data.py` and used them to verify that switching from `gamma`s in the denominator to using `rgamma` had an essentially negligible impact on accuracy.

Table here

| branch   |   min |         10% |         20% |         30% |         40% |         50% |         60% |         70% |         80% |         90% |         max |
|----------|-------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
| main     |     0 | 1.55481e-16 | 3.18565e-16 | 5.25405e-16 | 8.06051e-16 | 1.2675e-15  | 2.24815e-15 | 5.18608e-15 | 2.7321e-14  | 2.808e-12   | 2.01644e+44 |
| PR       |     0 | 1.54495e-16 | 3.15297e-16 | 5.22697e-16 | 8.04563e-16 | 1.26771e-15 | 2.24511e-15 | 5.20159e-15 | 2.72448e-14 | 2.82081e-12 | 2.01644e+44 |

this is over 3,003,855 test cases, covering all special cases.

h-vetinari commented:
> so we have a situation where we advise those using `scipy>=1.15` to use `rgamma` in gamma ratios, but for `scipy<1.15` using `rgamma` may be unacceptably inaccurate. @h-vetinari, do you have any thoughts on this?

I don't think the relative accuracy of `1e-13` that you showed is "unacceptably inaccurate". The reason I think this is not an issue is that it depends on someone sensitive to both:
* incorrectly getting Infs from poles of the gamma function
* requiring extreme accuracy that `rgamma` before this PR doesn't provide

I don't think many such cases exist, and if they do, they can write a version-dependent work-around, or just stay with `gamma` universally and do their own Inf/NaN-handling.

> I have extensive test cases for complex `hyp2f1` through `scipy/special/_precompute/hyp2f1_data.py` and used them to verify that switching from `gamma`s in the denominator to using `rgamma` had an essentially negligible impact on accuracy.

That further strengthens the argument above. :)

steppi commented:
Thanks @h-vetinari. I agree and am happy fixing this without deprecating the buggy behavior. It should be ready now

h-vetinari reviewed (approved):
Just made another pass and this LGTM! :)

Thank you very much for the diligence on this issue (& PR). I guess the one follow-up we could open is to investigate and possibly improve the "eyeballed" cutoff of 4 for the switch over in `rgamma` a bit more.

steppi commented:
Thanks @h-vetinari. Feel free to merge. I'll review https://github.com/scipy/scipy/pull/21801 after you update it to remove the now unnecessary workaround.

steppi commented on scipy/special/_special_ufuncs_docs.cpp:
```diff
@@ -1572,10 +1572,35 @@ const char *gamma_doc = R"(
+
+    Prior to SciPy version 1.15, ``scipy.special.gamma(z)`` returned ``+inf``
+    at each pole. This was fixed in version 1.15, but with the following
+    consequence. Expressions where gamma appears in the denominator of an
+    expression such as
```
```suggestion
    consequence. Expressions where gamma appears in the denominator
    such as
```

steppi committed:
Update scipy/special/_special_ufuncs_docs.cpp [docs only]

h-vetinari commented:
> Feel free to merge.

Did you want to maintain the full commit history here (already nicely separated, but quite a few), or would you prefer to squash? I'm fine either way.

steppi commented:
> > Feel free to merge.
> 
> Did you want to maintain the full commit history here (already nicely separated, but quite a few), or would you prefer to squash? I'm fine either way.

Go ahead and squash

h-vetinari commented:
I just rebased #21801 and noticed some failures because the following is still inconsistent:

> * 1. If the argument is $\pm 0$, $\pm \infty$ is returned.
> * 2. If the argument is a negative integer, NaN is returned.

Why should 0 be special compared to all the other negative integers? The behaviour is qualitatively exactly the same

![image](https://github.com/user-attachments/assets/6c556385-f942-49d3-a365-dfa5a82f1236)

I get the argument that this follows the C standard, but as it stands, I'd still need a wrapper to turn infinities at 0 to NaNs for the factorial implementation...

steppi commented:
> Why should 0 be special compared to all the other negative integers? The behaviour is qualitatively exactly the same

That's a great question actually. It's because zero has a sign bit in IEEE-754 floats, so we have both `-0.0` and `+0.0`. The sign bit can be used this to signify the direction in which `0` is being approached, so `gamma(+0.0) == +inf` and `gamma(-0.0) == -inf`. You may have also noticed that `1/+0.0 == +inf` and `1/-0.0 == -inf` for IEEE-754 floats, the situation is just the same here. This starts to look very practical when you notice that a zero can appear somewhere due to underflow, and you get to preserve the sign through underflow.

I'd be surprised if you really need to turn infinities at 0 to NaNs for your factorial implementation, it seems like a case where you'd want to use the sign bit of zero.

For added fun, Kahan has a really cool paper on the sign bit of zero [Much Ado about Nothing's Sign Bit](https://people.freebsd.org/~das/kahan86branch.pdf), particularly, about using it handle branch cuts in the complex plane.

steppi commented:
> I'd be surprised if you really need to turn infinities at 0 to NaNs for your factorial implementation, it seems like a case where you'd want to use the sign bit of zero.

Ah, I shouldn't have been surprised though. You have `factorial(-1) == gamma(0)`, and there's no "two-sidedness" for `-1`. Sorry, it looks like you still need a wrapper here. It might be more clean to make a private `_factorial` ufunc that's just `gamma(n + 1)`, special cases at `n = -1`.

h-vetinari commented:
> That's a great question actually. It's because zero has a sign bit in IEEE-754 floats, so we have both `-0.0` and `+0.0`.

I know that, and I consider `-0.0` to be an abomination. :)

> You have `factorial(-1) == gamma(0)`, and there's no "two-sidedness" for `-1`. Sorry, it looks like you still need a wrapper here.

Fine by me. 🙃

steppi commented:
@h-vetinari. Instead having a wrapper `gamma_nonconforming` or whatever, I think you should define a `gamma1p` that behaves in the natural way. It would naturally return `NaN` at -1.

Also if you  think `-0.0` is an abomination, I think the Kahan paper I linked above might change your mind. I used to be puzzled by signed zero too, but it's actually very useful.

h-vetinari commented:
> Instead having a wrapper `gamma_nonconforming` or whatever, I think you should define a private `gamma1p` that behaves in the natural way. It would naturally return `NaN` at -1. It can just wrap `gamma` and special case `x = -1`.

Would you like me to put this somewhere more prominent than `scipy/special/_basic.py`?

> Also if you think `-0.0` is an abomination, I think the Kahan paper I linked above might change your mind. I used to be puzzled by signed zero too, but it's actually very useful.

Thanks for the paper, I started reading it. So far I'm still convinced it's an abomination, though I'm willing to qualify that the signs make sense for complex math. That's where they should have been confined to. The damage `-0.0` _for real floats_ has done to general programming (in terms of bugs and wasted engineering time) dwarfs the benefits for the complex case IMO. Alas, it was too tempting to just assign the obvious bit pattern in general.

steppi commented:
> Actually, I think that module makes the most sense in any case, so putting it there now.

Yeah, agreed. It'll be a private function used only in `_basic.py` for now, so it seems like the best place to put it.

> Thanks for the paper, I started reading it. So far I'm still convinced it's an abomination, though I'm willing to qualify that the signs make sense for complex math. That's where they should have been confined to. The damage `-0.0` _for real floats_ has done to general programming (in terms of bugs and wasted engineering time) dwarfs the benefits for the complex case IMO. Alas, it was too tempting to just assign the obvious bit pattern in general.

It sounds like you have some war stories you could share :)  I like signed zeros. They have a nice symmetry with signed infinities. Signed infinities can represent $\infty$ and $-\infty$ in the extended real line, but also are the approximations for any real number that is too large to fit in the representable range. You also have an interval of real numbers $(-\epsilon, \epsilon)$ too small to fit in the representable range, and it's nice to be able to preserve at least their signs.
